### PR TITLE
Support for Extended Tasks

### LINKED ISSUE #837: Async reactive functions run sequentially instead of concurrently
In previous versions of Shiny, it did the following for async reactive functions:

- Within a session, async reactive functions _within_ ran sequentially. This was done to prevent a reactive flush from being run while another flush is happening.
- Across sessions, async functions ran concurrently. We decided this was safe, as long as there were no top-level async reactives which caused a flush in the middle of another flush.

It turns out that, as of now, both of these situations run async reactives sequentially. This is safe, but it also eliminates the performance benefits from async code.

This is a test app that demonstrates the issue. If you run the app and visit it with two browsers, and click on the button in each, you can see that the async `render.text`s run sequentially instead of concurrently.

```py
import asyncio
from shiny import App, Inputs, Outputs, Session, reactive, render, ui

app_ui = ui.page_fluid(
    ui.output_text("id"),
    ui.input_action_button("button", "Compute"),
    ui.output_text("compute1"),
)

def server(input: Inputs, output: Outputs, session: Session):
    # Get first 5 characters from Session
    id_num = session.id[:5]

    @render.text
    def id():
        return f"This session has ID: {id_num}"

    @render.text
    @reactive.event(input.button)
    async def compute1():
        for _ in range(8):
            print(f"... computing for ID {id_num}")
            await asyncio.sleep(0.5)
        return f"Finished run {input.button()} for ID {id_num}"

app = App(app_ui, server)
```

https://github.com/posit-dev/py-shiny/assets/86978/2eb987b7-3c0f-4df0-9f2c-a456332d2a9a

However, if we go into `shiny/session/_session.py` and comment out the following `async with lock`, then the two can run concurrently:
https://github.com/posit-dev/py-shiny/blob/351598bc41b1eec3d2068b7f75ec874b76eca498/shiny/session/_session.py#L272

https://github.com/posit-dev/py-shiny/assets/86978/c389da5f-7372-4d6d-b201-78b8187552cc

That `async with lock` was added in [4a0e5bff](https://github.com/posit-dev/py-shiny/commit/4a0e5bff7ff4a9eacffe5fcb8448d1265782b8fa#diff-593b4caeaea701b19634390e87e303e7003164d20d3afa0f773e085f3e0d39d1R195).

### PR DESCRIPTION
**Pull request by jcheng5**
This PR adds a couple of new features (that we're also adding to Shiny for R):

1. A `ui.input_task_button` that is similar to `ui.input_action_button`, except it automatically disables itself and shows "Processing..." after it clicks; and automatically(-ish) resets itself back to normal when the work it causes is done.
2. An `@extended_task` decorator that can be used to take a long-running task (expressed using an async function) and let it continue without blocking any reactive processing Shiny wants to do; but also makes it possible to take the results of such a computation and seamlessly integrate it back into reactive contexts, so it can be used from reactive outputs, calcs, etc.

Fixes #837 (or at least attempts to capture the spirit of it).

### CODE REVIEW

**Comment by jcheng5** [comment]
@schloerke @karangattu Either of you have any idea of why the CI playwright-examples tasks are failing?

A lot of it seems to be "UserWarning: Tight layout not applied. The left and right margins cannot be made large enough to accommodate all axes decorations." being emitted, but I don't see a way to know which particular example apps it was running when that happened.

**Comment by karangattu** [comment]
> @schloerke @karangattu Either of you have any idea of why the CI playwright-examples tasks are failing?
> 
> A lot of it seems to be "UserWarning: Tight layout not applied. The left and right margins cannot be made large enough to accommodate all axes decorations." being emitted, but I don't see a way to know which particular example apps it was running when that happened.

Update your branch...we made a few changes to example tests. I suspect your branch does not have those changes yet.

**Comment by schloerke** [comment]
@karangattu I believe it has the latest given the GHA runs on a merged commit. 

Looks like one error is coming from the cli create test. And the other is from layout_columns api example.

File: shiny/api-examples/extended_task/app.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+
+with ui.layout_sidebar():
+    with ui.sidebar():
+        ui.input_numeric("x", "x", 1)
+        ui.input_numeric("y", "y", 2)
+        ui.input_task_button("btn", "Compute, slowly")
+        ui.input_action_button("btn_cancel", "Cancel")
+
+    slow_compute.bind_task_button("btn")
```
Comment (gadenbuie) [inline shiny/api-examples/extended_task/app.py:33]: Alternatively (see other comment about where `bind_task_button()` lives)

```suggestion
    ui.bind_task_button("btn", slow_compute)
```

**Review by gadenbuie — APPROVED** [review]
This looks great! All of my comments are rather minor, except for one detail about whether `bind_task_button()` should be a method on `ExtendedTask` (which is a little different from how this ended up in Shiny for R).

**Review by gadenbuie — APPROVED** [review]

File: CHANGELOG.md
[the remark this answers is not in the document]
Comment (jcheng5) [reply CHANGELOG.md:1]: Ugggh I keep hitting this! Thank you

File: shiny/reactive/_extended_task.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        elif self.status() == "running":
+            req(False, cancel_output="progress")
+            # Will never get here, but make type checker happy
+            return self.value.get()
+        else:  # init, cancelled
+            req(False)
+            # Will never get here, but make type checker happy
+            return self.value.get()
+
+    def bind_task_button(self, id: str) -> None:
```
[the remark this answers is not in the document]
Comment (jcheng5) [reply shiny/reactive/_extended_task.py:204]: Yeah, I realize the approach I took here is not consistent with what we decided for R. I thought this was a little different for two reasons:

1. In R, we have separate bslib and shiny packages which made the distance between the binding logic and the button it's coupled to, feel more egregious.
2. In Python, `bind_task_button(task, "btn")` feels very weird for some reason.

That being said, I don't love `task.bind_task_button()` either. I was debating adding a `task_button_id` argument to the decorator. I realize this is moving even further in the opposite direction of what we decided for R.

[the remark this answers is not in the document]
Comment (gadenbuie) [reply shiny/reactive/_extended_task.py:204]: Reacting to `bind_task_button(task, "btn")` feeling weird: maybe binding to the task could be done through an additional decorator that applies to the ExtendedTask?

```suggestion
@ui.bind_task_button("btn")
@reactive.extended_task
async def slow_compute(a: int, b: int) -> int:
    await asyncio.sleep(3)
    return a + b
```

File: js/build.ts [hunk 3]
```diff
@@ -61,6 +61,11 @@ const opts: Array<BuildOptions> = [
     minify: false,
     sourcemap: false,
   },
+  {
+    entryPoints: { "spin/spin": "spin/spin.scss" },
+    plugins: [sassPlugin({ type: "css", sourceMap: false })],
```
[the remark this answers is not in the document]
Comment (jcheng5) [reply js/build.ts:66]: It's generated regardless. It doesn't look to me like `sourceMap` is a real option, not sure where it came from--I can remove it in both places where it's used.

[the remark this answers is not in the document]
Comment (jcheng5) [reply js/build.ts:66]: Hmmm, scratch that. It matters for the dataframe entry point; without `sourceMap: false` the source map is actually inlined into the the generated dataframe.js. With it, the .map file is generated but the inline sourcemap is gone.

File: shiny/reactive/_extended_task.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jcheng5) [reply shiny/reactive/_extended_task.py:200]: That's not bad. Let's look at it side by side with this:
```python
@reactive.extended_task(task_button_id="btn")
async def slow_compute(a: int, b: int) -> int:
    await asyncio.sleep(3)
    return a + b
```

[the remark this answers is not in the document]
Comment (gadenbuie) [reply shiny/reactive/_extended_task.py:200]: Personally, I'm pretty attached to the clean separation between an ExtendedTask and how we manifest the task state in the UI. Especially if we thought of a (hypothetical) alternative presentation like

```python
@ui.bind_task_modal()
@reactive.extended_task
async def slow_compute(a: int, b: int) -> int:
    await asyncio.sleep(3)
    return a + b
```

[the remark this answers is not in the document]
Comment (jcheng5) [reply shiny/reactive/_extended_task.py]: I tried it this way and didn't have a strong preference either way, so I went with it. To get it to work correctly as both function and decorator, the `id` argument has to be keyword-only. So the final formulation looks like:

```python
@ui.bind_task_button(button_id="btn")
@reactive.extended_task
async def slow_compute(a: int, b: int) -> int:
    await asyncio.sleep(3)
    return a + b
```

**Comment by jcheng5** [comment]
@schloerke @karangattu Turns out these playwright failures were not a false positive, but correctly calling out broken behavior from when I updated bslib!

@gadenbuie, when I created [bslib@v0.6.1-plus-task-button](https://github.com/rstudio/bslib/tree/v0.6.1-plus-task-button), I rebased my task button changes against the v0.6.1 CRAN release. It looks like that was too far back, as the Python `layout_columns` feature seems to rely on your refactoring in bslib that made layout_columns implemented by a web component. Can you help me figure out exactly what commits I need from bslib main, to be compatible with py-shiny main?

File: scripts/htmlDependencies.R [hunk 4]
```diff
@@ -11,7 +11,7 @@ versions <- list()
 message("Installing GitHub packages: bslib, shiny, htmltools")
 withr::local_temp_libpaths()
 ignore <- capture.output({
-  pak::pkg_install(c("rstudio/bslib", "cran::shiny", "cran::htmltools"))
+  pak::pkg_install(c("rstudio/bslib@v0.6.1-plus-task-button", "rstudio/shiny@main", "cran::htmltools"))
```
Comment (gadenbuie) [inline scripts/htmlDependencies.R:14]: Sorry about missing this! I think we should move away from allowing `@main` to be the ref here. Instead, if we're not using the version on CRAN, I propose we have a moving `py-shiny` tag in these repos and that we update those tags and before running this script.

Note that the actual ref used is stored in a JSON file when the dependencies are updated.

```suggestion
  pak::pkg_install(c("rstudio/bslib@py-shiny", "rstudio/shiny@py-shiny", "cran::htmltools"))
```

I'll add a tag for shiny and bslib. I'll point shiny at HEAD on main and I'll point bslib's tag at the previously used version of bslib so you can start from there.

File: shiny/www/shared/bslib/_version.json [hunk 5]
```diff
@@ -1,5 +1,5 @@
 {
   "note!": "This file is auto-generated by scripts/htmlDependencies.R",
   "package": "bslib",
-  "version": "Github (rstudio/bslib@f05bd23d7df5a7465e418a5794925dacdd27bb6b)"
```
Comment (gadenbuie) [inline shiny/www/shared/bslib/_version.json:4]: This is version of bslib we previously used in the last released version of py-shiny.

Comment (gadenbuie) [reply shiny/www/shared/bslib/_version.json:4]: Now tagged at https://github.com/rstudio/bslib/releases/tag/py-shiny

File: shiny/www/shared/_version.json [hunk 6]
```diff
@@ -1,5 +1,5 @@
 {
   "note!": "This file is auto-generated by scripts/htmlDependencies.R",
   "package": "shiny",
-  "version": "CRAN (R 4.3.1)"
+  "version": "Github (rstudio/shiny@54988c17c8fae0e74e69b3b7fd99edaa435dc3c6)"
```
Comment (gadenbuie) [inline shiny/www/shared/_version.json:4]: This is now tagged at https://github.com/rstudio/shiny/releases/tag/py-shiny

File: scripts/htmlDependencies.R (hunk 4 above)
Comment (gadenbuie) [reply scripts/htmlDependencies.R:14]: I guess there's a chicken-and-egg problem here: If we update the tag in bslib before merging in py-shiny, then for a period the tag is out of date. If we want to be careful, as in this PR, we can start the PR with a branch or short ref and then, before merging move the tag and update this script to use the new tag.

Comment (jcheng5) [reply scripts/htmlDependencies.R]: I'm taking the careful approach. 👍

**Pull request merged**